File: apps/web/src/lib/scraper/fetcher.ts
export interface FetchResult {
  url: string
  html: string
  status: number
  contentType: string
  fetchedAt: Date
  error?: string
}

const DEFAULT_HEADERS: Record<string, string> = {
  'User-Agent':
    'Mozilla/5.0 (compatible; ToyIntelBot/1.0; +https://toyintel.com/bot)',
  'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
  'Accept-Language': 'en-US,en;q=0.5',
  'Accept-Encoding': 'gzip, deflate, br',
  'Cache-Control': 'no-cache',
}

const TIMEOUT_MS = 30000

export async function fetchPage(url: string): Promise<FetchResult> {
  const controller = new AbortController()
  const timeout = setTimeout(() => controller.abort(), TIMEOUT_MS)

  try {
    const response = await fetch(url, {
      headers: DEFAULT_HEADERS,
      signal: controller.signal,
      redirect: 'follow',
    })

    const contentType = response.headers.get('content-type') || ''
    const html = await response.text()

    return {
      url,
      html,
      status: response.status,
      contentType,
      fetchedAt: new Date(),
    }
  } catch (err) {
    if (err instanceof Error && err.name === 'AbortError') {
      return {
        url,
        html: '',
        status: 0,
        contentType: '',
        fetchedAt: new Date(),
        error: `Request timed out after ${TIMEOUT_MS}ms`,
      }
    }
    return {
      url,
      html: '',
      status: 0,
      contentType: '',
      fetchedAt: new Date(),
      error: err instanceof Error ? err.message : String(err),
    }
  } finally {
    clearTimeout(timeout)
  }
}

export async function fetchRSS(url: string): Promise<FetchResult> {
  const result = await fetchPage(url)
  return result
}

export function isSuccessfulFetch(result: FetchResult): boolean {
  return result.status >= 200 && result.status < 300 && !result.error && result.html.length > 0
}

export function extractTextContent(html: string): string {
  // Remove script and style tags
  let text = html
    .replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '')
    .replace(/<style\b[^<]*(?:(?!<\/style>)<[^<]*)*<\/style>/gi, '')
    // Remove HTML tags
    .replace(/<[^>]+>/g, ' ')
    // Normalize whitespace
    .replace(/\s+/g, ' ')
    .trim()

  return text
}

File: services/scraper/fetcher.py
"""HTTP fetcher with Playwright fallback for JS-heavy pages."""

import asyncio
from dataclasses import dataclass, field
from typing import Optional
from datetime import datetime

import httpx
from loguru import logger
from config import config

DEFAULT_HEADERS = {
    "User-Agent": "Mozilla/5.0 (compatible; ToyIntelBot/1.0; +https://toyintel.com/bot)",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    "Accept-Language": "en-US,en;q=0.5",
    "Cache-Control": "no-cache",
}


@dataclass
class FetchResult:
    url: str
    html: str
    status: int
    content_type: str
    fetched_at: datetime = field(default_factory=datetime.utcnow)
    error: Optional[str] = None
    used_playwright: bool = False


async def fetch_with_httpx(url: str) -> FetchResult:
    """Fetch a page using httpx (faster, no JS rendering)."""
    try:
        async with httpx.AsyncClient(
            headers=DEFAULT_HEADERS,
            timeout=config.FETCH_TIMEOUT,
            follow_redirects=True,
        ) as client:
            response = await client.get(url)
            return FetchResult(
                url=url,
                html=response.text,
                status=response.status_code,
                content_type=response.headers.get("content-type", ""),
            )
    except Exception as e:
        logger.warning(f"httpx fetch failed for {url}: {e}")
        return FetchResult(url=url, html="", status=0, content_type="", error=str(e))


async def fetch_with_playwright(url: str) -> FetchResult:
    """Fetch a page using Playwright (handles JS rendering)."""
    try:
        from playwright.async_api import async_playwright

        async with async_playwright() as p:
            browser = await p.chromium.launch(headless=True)
            context = await browser.new_context(
                user_agent=DEFAULT_HEADERS["User-Agent"],
                extra_http_headers={"Accept-Language": "en-US,en;q=0.5"},
            )
            page = await context.new_page()
            await page.goto(url, wait_until="networkidle", timeout=config.FETCH_TIMEOUT * 1000)
            html = await page.content()
            status = 200  # Playwright doesn't expose status easily after navigation
            await browser.close()

            return FetchResult(
                url=url,
                html=html,
                status=status,
                content_type="text/html",
                used_playwright=True,
            )
    except Exception as e:
        logger.warning(f"Playwright fetch failed for {url}: {e}")
        return FetchResult(
            url=url, html="", status=0, content_type="", error=str(e), used_playwright=True
        )


async def fetch_page(url: str, use_playwright: bool = False) -> FetchResult:
    """Fetch a page, using Playwright if requested or if httpx fails with JS hint."""
    if use_playwright:
        return await fetch_with_playwright(url)

    result = await fetch_with_httpx(url)

    # Fallback to Playwright if httpx returns nearly empty content
    if result.status == 200 and len(result.html) < 1000:
        logger.info(f"httpx returned small content for {url}, trying Playwright")
        pw_result = await fetch_with_playwright(url)
        if len(pw_result.html) > len(result.html):
            return pw_result

    return result


def is_successful(result: FetchResult) -> bool:
    return result.status >= 200 and result.status < 300 and not result.error and len(result.html) > 0


async def fetch_rss(url: str) -> FetchResult:
    """Fetch an RSS/Atom feed."""
    return await fetch_with_httpx(url)
